aws-cpp-sdk-transfer/source/transfer/UploadFileRequest.cpp

```diff
@@ -577,27 +546,35 @@ bool UploadFileRequest::RequestPart(uint32_t partId)
     PartRequestRecord& partRequest = partIter->second;
 
     partRequest.m_retries++;
-    std::shared_ptr<Aws::Client::AsyncCallerContext> context = Aws::MakeShared<UploadFileContext>(ALLOCATION_TAG, shared_from_this());
+    
+    thisLock.unlock();
 
-    GetS3Client()->UploadPartAsync(partRequest.m_partRequest, &TransferClient::OnUploadPartRequest, context);
+    m_executor->Submit(&UploadFileRequest::UploadPartHelper, this, partRequest.m_partRequest);
 
     return true;
 }
+    
+void UploadFileRequest::UploadPartHelper(const Aws::S3::Model::UploadPartRequest& request)
+{
+    auto outcome = GetS3Client()->UploadPart(request);
+    HandleUploadPartOutcome(request, outcome);
+}
 
 bool UploadFileRequest::HandleUploadPartOutcome(const Aws::S3::Model::UploadPartRequest& request, const Aws::S3::Model::UploadPartOutcome& outcome)
 {
     Aws::StringStream md5Hex;
 
     PartRequestRecord partRequest;
     {
-        std::lock_guard<std::mutex> pendingLock(m_pendingMutex);
+        std::unique_lock<std::mutex> pendingLock(m_pendingMutex);
         auto partIter = m_pendingParts.find(request.GetPartNumber());
 
         if (partIter == m_pendingParts.end())
         {
             CompletionFailure("Bad part returned");
             return false;
         }
+        pendingLock.unlock();

```

Comment:
Because I don't know this code, it's really hard for me to tell if these locking changes are correct, i.e., can you really look at partIter after you unlock?